File: node/packages/aws-lambda-sdk/lib/instrumentation/aws-sdk/service-mapper.js
'use strict';

const dynamodbConfig = {
  params: (
    traceSpan,
    {
      TableName: tableName,
      GlobalTableName: globalTableName,
      ConsistentRead: consistentRead,
      Limit: limit,
      AttributesToGet: attributesToGet,
      ProjectionExpression: projectionExpression,
      IndexName: indexName,
      ScanIndexForward: scanIndexForward,
      Select: select,
      KeyConditionExpression: keyConditionExpression,
      FilterExpression: filterExpression,
      Segment: segment,
      TotalSegments: totalSegments,
      ExclusiveStartKey: exclusiveStartKey,
    }
  ) => {
    const tags = {};
    if (tableName) tags.table_name = tableName;
    else if (globalTableName) tags.table_name = globalTableName;
    if (consistentRead) tags.consistent_read = consistentRead;
    if (limit) tags.limit = limit;
    tags.attributes_to_get = attributesToGet || [];
    if (projectionExpression) tags.projection = projectionExpression;
    if (indexName) tags.index_name = indexName;
    if (scanIndexForward) tags.scan_forward = scanIndexForward;
    if (select) tags.select = select;
    if (filterExpression) tags.filter = filterExpression;
    if (keyConditionExpression) tags.key_condition = keyConditionExpression;
    if (segment) tags.segment = segment;
    if (totalSegments) tags.total_segments = totalSegments;
    if (exclusiveStartKey) tags.exclusive_start_key = JSON.stringify(exclusiveStartKey);
    traceSpan.tags.setMany(tags, { prefix: 'aws.sdk.dynamodb' });
  },
  responseData: (traceSpan, { Count: count, ScannedCount: scannedCount }) => {
    const tags = {};
    if (count) tags.count = count;
    if (scannedCount) tags.scanned_count = scannedCount;
    traceSpan.tags.setMany(tags, { prefix: 'aws.sdk.dynamodb' });
  },
};

module.exports = new Map([
  [
    'sns',
    {
      params: (traceSpan, { TopicArn: topicArn }) => {
        const tags = { message_ids: [] };
        if (topicArn) tags.topic_name = topicArn.slice(topicArn.lastIndexOf(':') + 1);
        traceSpan.tags.setMany(tags, { prefix: 'aws.sdk.sns' });
      },
      responseData: (
        traceSpan,
        { TopicArn: topicArn, MessageId: messageId, Successful: messages }
      ) => {
        traceSpan.tags.delete('aws.sdk.sns.message_ids');
        const tags = {};
        if (topicArn) tags.topic_name = topicArn.slice(topicArn.lastIndexOf(':') + 1);
        if (messageId) tags.message_ids = [messageId];
        else if (messages) tags.message_ids = messages.map(({ MessageId: id }) => id);
        else tags.message_ids = [];
        traceSpan.tags.setMany(tags, { prefix: 'aws.sdk.sns' });
      },
    },
  ],
  [
    'sqs',
    {
      params: (traceSpan, { QueueUrl: queueUrl, QueueName: queueName }) => {
        const tags = { message_ids: [] };
        if (queueUrl) tags.queue_name = queueUrl.slice(queueUrl.lastIndexOf('/') + 1);
        else if (queueName) tags.queue_name = queueName;
        traceSpan.tags.setMany(tags, { prefix: 'aws.sdk.sqs' });
      },
      responseData: (
        traceSpan,
        {
          QueueUrl: queueUrl,
          MessageId: messageId,
          Successful: successfulMessages,
          Messages: messages,
        }
      ) => {
        traceSpan.tags.delete('aws.sdk.sqs.message_ids');
        const tags = {};
        if (queueUrl) tags.queue_name = queueUrl.slice(queueUrl.lastIndexOf('/') + 1);
        if (messageId) {
          tags.message_ids = [messageId];
        } else if (successfulMessages || messages) {
          tags.message_ids = (successfulMessages || messages)
            .map(({ MessageId: id }) => id)
            .filter(Boolean);
        } else {
          tags.message_ids = [];
        }
        traceSpan.tags.setMany(tags, { prefix: 'aws.sdk.sqs' });
      },
    },
  ],
  ['dynamodb', dynamodbConfig],
  ['dynamodbdocument', dynamodbConfig],
]);

File: python/packages/aws-lambda-sdk/serverless_aws_lambda_sdk/lib/instrumentation/aws_sdk/service_mapper.py
from abc import ABC, abstractmethod


class ServiceMapper(ABC):
    @abstractmethod
    def params(self, trace_span, input):
        tags = {}
        if "TableName" in input:
            tags["table_name"] = input["TableName"]
        elif "GlobalTableName" in input:
            tags["table_name"] = input["GlobalTableName"]
        if "ConsistentRead" in input:
            tags["consistent_read"] = input["ConsistentRead"]
        if "Limit" in input:
            tags["limit"] = input["Limit"]
        tags["attributes_to_get"] = input.get("AttributesToGet", [])
        if "ProjectionExpression" in input:
            tags["projection"] = input["ProjectionExpression"]
        if "IndexName" in input:
            tags["index_name"] = input["IndexName"]
        if "ScanIndexForward" in input:
            tags["scan_forward"] = input["ScanIndexForward"]
        if "Select" in input:
            tags["select"] = input["Select"]
        if "KeyConditionExpression" in input:
            tags["key_condition"] = input["KeyConditionExpression"]
        if "FilterExpression" in input:
            tags["filter"] = input["FilterExpression"]
        if "Segment" in input:
            tags["segment"] = input["Segment"]
        if "TotalSegments" in input:
            tags["total_segments"] = input["TotalSegments"]
        if "ExclusiveStartKey" in input:
            tags["exclusive_start_key"] = input["ExclusiveStartKey"]
        trace_span.tags.update(tags, "aws.sdk.dynamodb")

    @abstractmethod
    def response_data(self, trace_span, response):
        tags = {}
        if "Count" in response:
            tags["count"] = response["Count"]
        if "ScannedCount" in response:
            tags["scanned_count"] = response["ScannedCount"]
        trace_span.tags.update(tags, "aws.sdk.dynamodb")


class DynamoDBMapper(ServiceMapper):
    def params(self, trace_span, input):
        trace_span.tags.set("aws.sdk.table_name", input.get("TableName", ""))

    def response_data(self, trace_span, response):
        trace_span.tags.set("aws.sdk.item_count", len(response.get("Items", [])))


class SQSMapper(ServiceMapper):
    def params(self, trace_span, input):
        tags = {
            "message_ids": [],
        }
        if "QueueUrl" in input:
            tags["queue_name"] = input.get("QueueUrl").split("/")[-1]
        elif "QueueName" in input:
            tags["queue_name"] = input.get("QueueName")
        trace_span.tags.update(tags, "aws.sdk.sqs")

    def response_data(self, trace_span, response):
        del trace_span.tags["aws.sdk.sqs.message_ids"]
        tags = {}
        if "QueueUrl" in response:
            tags["queue_name"] = response.get("QueueUrl").split("/")[-1]
        if "MessageId" in response:
            tags["message_ids"] = [response.get("MessageId")]
        elif "Successful" in response or "Messages" in response:
            tags["message_ids"] = [
                message.get("MessageId")
                for message in response.get("Successful", response.get("Messages", []))
                if message.get("MessageId")
            ]
        else:
            tags["message_ids"] = []
        trace_span.tags.update(tags, "aws.sdk.sqs")


class SNSMapper(ServiceMapper):
    def params(self, trace_span, input):
        tags = {
            "message_ids": [],
        }
        if "TopicArn" in input:
            tags["topic_name"] = input.get("TopicArn").split(":")[-1]
        trace_span.tags.update(tags, "aws.sdk.sns")

    def response_data(self, trace_span, response):
        del trace_span.tags["aws.sdk.sns.message_ids"]
        tags = {}
        if "TopicArn" in response:
            tags["topic_name"] = response.get("TopicArn").split(":")[-1]
        if "MessageId" in response:
            tags["message_ids"] = [response.get("MessageId")]
        elif "Successful" in response:
            tags["message_ids"] = [
                message.get("MessageId") for message in response.get("Successful", [])
            ]
        trace_span.tags.update(tags, "aws.sdk.sns")


_SERVICE_MAPPERS = {
    "dynamodb": DynamoDBMapper(),
    "sqs": SQSMapper(),
    "sns": SNSMapper(),
}


def get_mapper_for_service(service_name):
    return _SERVICE_MAPPERS[service_name]
